cos_alerter/alerter.py

```diff
@@ -29,24 +34,63 @@ def __exit__(self, _, __, ___):
         fcntl.lockf(self.fh, fcntl.LOCK_UN)
         self.fh.close()
 
-    @property
-    def alert_time(self):
-        return self.data['alert_time']
+    @staticmethod
+    def initialize():
+        # Note this method does not do any locking so be responsible.
+        current_date = datetime.datetime.now(datetime.timezone.utc)
+        current_time = time.monotonic()
+        data = {
+            # The actual date and time that COS Alerter was started.
+            'start_date': datetime.datetime.timestamp(current_date),
 
-    @alert_time.setter
-    def alert_time(self, value):
-        self.data['alert_time'] = value
+            # The time according to the monotonic clock that COS Alerter was started.
+            'start_time': current_time,
 
-    @property
-    def notify_time(self):
-        return self.data['notify_time']
+            # The last time we received an alert from Alertmanager.
+            'alert_time': current_time,
 
-    @notify_time.setter
-    def notify_time(self, value):
-        self.data['notify_time'] = value
+            # The last time we sent out notifications.
+            'notify_time': None,
+        }
+
+        with open(config['watch']['data_file'], 'w') as f:
+            json.dump(data, f)
+
+    def set_alert_time(self):
+        self.data['alert_time'] = time.monotonic()
+
+    def _set_notify_time(self):
+        self.data['notify_time'] = time.monotonic()
+
+    def is_down(self):
+        down_interval = durationpy.from_str(config['watch']['down_interval']).total_seconds()
+        if time.monotonic() - self.data['alert_time'] > down_interval:
+            return True
+        return False
+
+    def notify(self):
+        # If we have already notified recently, do nothing.
+        repeat_interval = durationpy.from_str(config['notify']['repeat_interval']).total_seconds()
+        if self.data['notify_time'] and not time.monotonic() - self.data['notify_time'] > repeat_interval:
+            return
+
+        self._set_notify_time()
+        actual_alert_timestamp = (self.data['alert_time'] - self.data['start_time']) + self.data['start_date']
+        last_alert_time = datetime.datetime.fromtimestamp(actual_alert_timestamp, datetime.timezone.utc).isoformat()
+        title = '**Alertmanager is Down!**'
+        body = textwrap.dedent(f'''
+            Your Alertmanager instance seems to be down!
+            It has not alerted COS-Alerter since {last_alert_time} UTC.
+            ''')
+
+        # Sending notifications can be a long operation so handle that in a separate thread.
+        notify_thread = threading.Thread(target=send_notifications, kwargs={'title': title, 'body': body})
+        notify_thread.start()
```

Comment:
**question:** Why do we need a thread here?

This is a thread, not a fork, to charm code won't exit anyway.
And because charm code won't exit anyway, then juju lease timeout (for killing or re-electing a leader) may still equally apply with or without a thread.

Reply:
This won't be running in Juju or as a charm in eithe rcas.

The thread is potentially nice due to the amount of time it may take to send an email/webhook, that it could be joined and re-started if there's a signal, and arguably getting its own stack is nice because flow control from exceptions (or joining/restarting the thread on SIGHUP or something) could be done semi-independently.

Reply:
> This won't be running in Juju or as a charm in eithe rcas.

Right!

> it could be joined and re-started if there's a signal
> flow control from exceptions could be done semi-independently

I'm not sure I understand the use case.
1. How long does it take? Does it matter if it's 5 or 10 sec?
2. Why the amount of time it takes matters at all / prompts for a thread?

Reply:
- Makes sense. Summarize this in the comment?
- Does `send_notifications` have some kind of a timeout? If not, and it hangs, would we end up with multiple threads?
- Do we need to enforce that the timeout is shorter than `repeat_interval`?

Reply:
* [Improved the comment] in the branch for my next PR.
* Good point. I will check.
* I think probably not. If repeat interval is less than an email timeout, you probably don't mind getting two in a row.

Reply:
Issue [here](https://github.com/caronc/apprise-api/issues/112) I think we should leave it for now and open a new issue if there is no timeout.

Reply:
There are global params available in all URLs to override the default timeout:

> cto | This stands for Socket Connect Timeout. This is the number of seconds Requests 
 will wait for your client to establish a connection to a remote machine (corresponding to the connect()) call on the socket. The default value is 4.0 seconds.
> rto | This stands for Socket Read Timeout. This is the number of seconds the client will wait for the server to send a response. The default value is 4.0 seconds.

https://github.com/caronc/apprise/wiki/URLBasics#global-parameter-breakdown